test/sys/test_socket.rs

```diff
@@ -2828,3 +2828,87 @@ fn test_icmp_protocol() {
     sendto(owned_fd.as_raw_fd(), &packet, &dest_addr, MsgFlags::empty())
         .unwrap();
 }
+
+#[cfg(any(apple_targets, linux_android))]
+#[test]
+pub fn test_recv_iptos_ipttl() {
+    use nix::sys::socket::sockopt::{IpRecvTos, IpRecvTtl}; //, IpTos, IpTtl};
+    use nix::sys::socket::{bind, SockFlag, SockType, SockaddrIn};
+    use nix::sys::socket::{getsockname, setsockopt, socket};
+    use nix::sys::socket::{recvmsg, sendmsg, ControlMessageOwned, MsgFlags};
+    use std::io::{IoSlice, IoSliceMut};
+
+    let lo_ifaddr = loopback_address(AddressFamily::Inet);
+    let (_, lo) = match lo_ifaddr {
+        Some(ifaddr) => (
+            ifaddr.interface_name,
+            ifaddr.address.expect("Expect IPv4 address on interface"),
+        ),
+        None => return,
+    };
+    let receive = socket(
+        AddressFamily::Inet,
+        SockType::Datagram,
+        SockFlag::empty(),
+        None,
+    )
+    .expect("receive socket failed");
+    bind(receive.as_raw_fd(), &lo).expect("bind failed");
+    let sa: SockaddrIn =
+        getsockname(receive.as_raw_fd()).expect("getsockname failed");
+    setsockopt(&receive, IpRecvTos, &true).expect("setsockopt failed");
+    setsockopt(&receive, IpRecvTtl, &true).expect("setsockopt failed");
+
+    {
+        let slice = [1u8, 2, 3, 4, 5, 6, 7, 8];
+        let iov = [IoSlice::new(&slice)];
+
+        let send = socket(
+            AddressFamily::Inet,
+            SockType::Datagram,
+            SockFlag::empty(),
+            None,
+        )
+        .expect("send socket failed");
+        sendmsg(send.as_raw_fd(), &iov, &[], MsgFlags::empty(), Some(&sa))
+            .expect("sendmsg failed");
+    }
+
+    {
+        let mut buf = [0u8; 8];
+        let mut iovec = [IoSliceMut::new(&mut buf)];
+
+        // FIXME: When passing in IpTos and IpTtl, cmsg_space returns too little space?
+        // let mut space = cmsg_space!(IpTos, IpTtl);
+        let mut space = cmsg_space!(libc::c_int, libc::c_int);
```

Comment:
@asomers I'm trying to look into this, but I am new to Rust. What I have figured out is that `mem::size_of::< nix::sys::socket::sockopt::IpTos>()` is `0`, and that causes the issue, because `cmsg_space!` allocates too little space. What I don't understand is _why_ it is `0`, given that I have (tried to) make it a `libc::c_int`...

Reply:
Also `test_recv_ipv4pktinfo` also allocates via `let mut space = cmsg_space!(libc::in_pktinfo)` and not via `let mut space = cmsg_space!(nix::sys::socket::sockopt::Ipv4PacketInfo)`, so I guess that's the way to do it?

Reply:
Here's the answer: you were trying to calculate the cmsg_space of something unrelated.  Instead of `nix::sys::socket::sockopt::IpTos`, you really wanted `nix::sys::socket::ControlMessageOwned::IpTos`.  The former really does have a size of 0.  And you can't pass the latter to `cmsg_space!`, because `mem::size_of` doesn't work on enum variants.  Does that explanation change anything about the way you would write your test?